src/internal/connector/graph/service.go

```diff
@@ -61,6 +78,90 @@ func (s Service) Serialize(object absser.Parsable) ([]byte, error) {
 	return writer.GetSerializedContent()
 }
 
+// ---------------------------------------------------------------------------
+// Helpers
+// ---------------------------------------------------------------------------
+
+type clientConfig struct {
+	noTimeout bool
+}
+
+type option func(*clientConfig)
+
+// populate constructs a clientConfig according to the provided options.
+func (c *clientConfig) populate(opts ...option) *clientConfig {
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c
+}
+
+// apply updates the http.Client with the expected options.
+func (c *clientConfig) apply(hc *http.Client) {
+	if c.noTimeout {
+		hc.Timeout = 0
+	}
+}
+
+// NoTimeout sets the httpClient.Timeout to 0 (unlimited).
+// The resulting client isn't suitable for most queries, due to the
+// capacity for a call to persist forever.  This configuration should
+// only be used when downloading very large files.
+func NoTimeout() option {
+	return func(c *clientConfig) {
+		c.noTimeout = true
+	}
+}
+
+// CreateAdapter uses provided credentials to log into M365 using Kiota Azure Library
+// with Azure identity package. An adapter object is a necessary to component
+// to create  *msgraphsdk.GraphServiceClient
+func CreateAdapter(tenant, client, secret string, opts ...option) (*msgraphsdk.GraphRequestAdapter, error) {
```

Comment:
I'm curious as to the choice of passing a vararg of function pointers rather than just a `clientConfig` struct that's then consumed by `HTTPClient()`. It seems like the logic would be easier to follow on the latter while still allowing the user to modify the full set of parameters. It may also be more flexible as we wouldn't be bound to functions that only take `*clientConfig` as a parameter (e.x. being able to set the timeout to any integer vs. having different functions for different integer values)

Reply:
> I'm curious as to the choice of passing a vararg of function pointers rather than just a clientConfig struct that's then consumed by HTTPClient().

I like this pattern because it makes the base (ie, no options) constructor look like it produces a viable default value without further input.  This keeps the majority of our cases the most simple.  Requiring a configuration struct adds chaff complexity to the majority of default construction.  Empty and nil configurations can be confusing, as the caller doesn't know at first glance if they're _expected to_ be configuring further.  Options read more naturally like like deviations from the default.

> (e.x. being able to set the timeout to any integer vs. having different functions for different integer values)

Oh, this is a very easy change.  `NoTimeout()` -> `Timeout(0)`.  These funcs don't need to be bools.

Reply:
>Oh, this is a very easy change. NoTimeout() -> Timeout(0). These funcs don't need to be bools.

Ah, I see. I still find that rather annoying to read because now there's multiple levels of indirection to parse through just to figure out what parameters are being passed. For example, we have function call for param -> functor -> populate function -> apply function and then the reader has to realize that it's actually the union of all populate  calls which is kind of difficult to hold in your head at once. Also things like `Timeout(0), Timeout(5)` become valid and then ordering is thrown in on top of everything else

It still seems like a simple `DefaultConfig()` would give known good values even if it adds a little bit of chaff

Reply:
> Also things like Timeout(0), Timeout(5) become valid and then ordering is thrown in on top of everything else

Sure, but that's also a bit contrived, don't you think?  Under a microscope, yes, I absolutely agree there's weirdness.  But that weirdness is escaped in real usage by duplicate configurations being obviously in contention.  Now, some of this is on us as curators, sure.  For example, `NoTimeout(), Timeout(N)`, is both very weird and less obvious.  But that can be avoided by ensuring that we have either Timeout() or NoTimeout(), and not both.

> I still find that rather annoying to read because now there's multiple levels of indirection to parse through just to figure out what parameters are being passed.

In the perspective of calling the constructor, or handling it?  Personally, I think it's a tradeoff where usage of the constructor gets more readable, with only a minor increase in complexity when processing construction.

We're both soapboxing a bit about design philosophy and preferred patterns here.  This is good, we want to pin down the patterns that we prefer (or prefer to avoid) within the code.  But I also feel like this is the wrong place to resolve the discussion.  Are you okay with treating this as a two-way door that we can come back to after we resolve code standards elsewhere?

Reply:
>Sure, but that's also a bit contrived, don't you think?

It depends. If we end up with lots of different settings then it can become easy to miss stuff like this because it's easy to skim over long lists of parameters or just think "I'll just add this at the end because it's not set already"

>In the perspective of calling the constructor, or handling it?

From the perspective of calling the constructor. Unless someone really reads the code well and is familiar with this pattern then it's something that's easy to miss or misunderstand (e.x. me thinking we couldn't have function parameters) which harms long-term maintainability of the codebase

>Are you okay with treating this as a two-way door that we can come back to after we resolve code standards elsewhere?

Only so long as we have a fixed date where we resolve this question and make time to actually go through the codebase and do cleanup. I don't want to get into the habit of saying we'll resolve things elsewhere and then never revisiting them

Reply:
> I don't want to get into the habit of saying we'll resolve things elsewhere and then never revisiting them

P2 tech debt, yeah? 😁

Reply:
🥲 RIP P2 tech debt. Approving for now as there's an out-of-band discussion for determining and laying down standards